implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake.rs

```diff
@@ -0,0 +1,641 @@
+        let (encryption_key, decryption_key) = if role.is_initiator() {
+            (k2, k1)
+        } else {
+            (k1, k2)
+        };
+        state.status = Ready(HandshakeKeys {
+            encryption_key,
+            decryption_key,
+        });
+
+        self.state = state;
+        Ok(())
+    }
+
+    /// Return the final results of the handshake if we reached the final state
+    pub(super) fn get_handshake_keys(&self) -> Option<HandshakeKeys> {
+        match self.state.status.clone() {
+            Ready(keys) => Some(keys),
+            _ => None,
+        }
+    }
+}
+
+impl Handshake {
+    /// Create a new handshake initialized with
+    ///   - a static key
+    ///   - an ephemeral key
+    ///   - a payload containing the identity, a signature of the static key and the credential of
+    ///     the current party
+    pub(super) async fn new(vault: Arc<dyn XXVault>) -> Result<Handshake> {
+        // 1. generate a static key pair for this handshake and set it to s
+        let static_key = Self::generate_static_key(vault.clone()).await?;
+
+        // 2. generate an ephemeral key pair for this handshake and set it to e
+        let ephemeral_key = Self::generate_ephemeral_key(vault.clone()).await?;
+
+        // 3. initialize the handshake
+        // We currently don't use any payload for message 1
+        Self::new_with_keys(vault, static_key, ephemeral_key, vec![]).await
+    }
+
+    /// Initialize the handshake
+    async fn new_with_keys(
+        vault: Arc<dyn XXVault>,
+        static_key: KeyId,
+        ephemeral_key: KeyId,
+        message1_payload: Vec<u8>,
+    ) -> Result<Handshake> {
+        Ok(Handshake {
+            vault,
+            state: HandshakeState::new(static_key, ephemeral_key, message1_payload),
+        })
+    }
+
+    /// Import the k secret
+    async fn import_k_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::k_attributes())
+            .await
+    }
+
+    /// Import the ck secret
+    async fn import_ck_secret(&self, content: Vec<u8>) -> Result<KeyId> {
+        self.vault
+            .import_ephemeral_secret(Secret::new(content), Self::ck_attributes())
+            .await
+    }
+
+    /// Return the public key corresponding to a given key id
+    async fn get_public_key(&self, key_id: &KeyId) -> Result<PublicKey> {
+        self.vault.get_public_key(key_id).await
+    }
+
+    /// Compute a Diffie-Hellman key between a given key id and the other party public key
+    async fn dh(&self, key_id: &KeyId, public_key: &PublicKey) -> Result<KeyId> {
+        self.vault.ec_diffie_hellman(key_id, public_key).await
+    }
+
+    /// Compute two derived ck, and k keys based on existing ck and k keys + an optional
+    /// Diffie-Hellman key
+    async fn hkdf(&self, state: &mut HandshakeState, dh: Option<&KeyId>) -> Result<(KeyId, KeyId)> {
```

Comment:
Does it make sense to split this function into 2 cases:
1. Derive new `ck` and `k`
2. Derive `k1` and `k2`
?

Reply:
Yes that's better. Done.